src/lib.rs

```
@@ -336,22 +340,97 @@ impl Config {
-            src_dst.push((file.to_path_buf(), obj.clone()));
+
+            let mut input_mtime = get_mtime(&file);
+            let output_mtime = get_mtime(&obj);
+
+            let mut cmd = compiler.to_command();
+            if msvc {
+                cmd.arg("/showIncludes").arg("/E");
+            } else {
+                cmd.arg("-M").arg("-MG");
+            }
+            cmd.arg(&file);
+            println!("running: {:?}", cmd);
+
+            let cpp_output = match cmd.output() {
+                Ok(ref out) if out.status.success() => {
+                    String::from_utf8(if msvc {
+                        out.stderr.clone()
+                    } else {
+                        out.stdout.clone()
+                    }).ok()
+                },
+                _ => {
+                    println!("warning: dependency check failed, building anyway: {:?}", file);
+                    input_mtime = output_mtime;
+                    None
+                }
+            };
+
+            if let Some(cpp) = cpp_output {
+                let files = if msvc {
+                    // Output is indented dependency per line, with "Note: ..." prefix 
+                    cpp
+                        .replace("Note: including file:", "")
+                } else {
+                    // Output is make rule with line continuations and multiple files per line
+                    cpp
+                        .replace("\\\n", "") // remove line continuations
+                        .replace("\\ ", "\0") // protect escaped spaces
+                        .replace(" ", "\n") // spaces to new lines
+                        .replace("\0", "\\ ") // bring escaped spaces back
+                };
+
+                input_mtime = files
+                    .lines().skip(1)
+                    .chain(vec![file.to_str().unwrap_or("")])
+                    .filter_map(|s| PathBuf::from(s.trim()).canonicalize().ok())
```

I'm also a little wary of ignoring all errors here. This means that if one header file has a problem then we'll ignore it even if it changed.